### PR TITLE
core, epochStart: economics invariants

### PR DESCRIPTION
**Pull request by AdoAdoAdo**
add some invariants for end of epoch economics

### CODE REVIEW

File: epochStart/metachain/economics.go [hunk 1]
```diff
@@ -185,6 +185,13 @@ func (e *economics) ComputeEndOfEpochEconomics(
 		rewardsForProtocolSustainability,
 	)
 
+	err = e.checkEconomicsInvariants(computedEconomics, inflationRate, maxBlocksInEpoch, totalNumBlocksInEpoch, metaBlock, metaBlock.Epoch)
+	if err != nil {
+		log.Warn("ComputeEndOfEpochEconomics", "error", err.Error())
+		// TODO: return maybe old epoch economics instead of failing?
```
Comment (iulianpascalau) [inline epochStart/metachain/economics.go:191]: Open for debate: we can return the old economics or a computed economics struct holding 0 rewards/accumulated fees and such.

File: epochStart/metachain/economics.go [hunk 2]
```diff
@@ -406,6 +419,54 @@ func (e *economics) startNoncePerShardFromLastCrossNotarized(metaNonce uint64, e
...
+	totalNumBlocksInEpoch uint64,
+	metaBlock *block.MetaBlock,
+	epoch uint32,
+) error {
+	if epoch <= e.stakingV2EnableEpoch {
+		return nil
+	}
+
+	if !core.IsInRangeInclusiveFloat64(inflationRate, 0, e.rewardsHandler.MaxInflationRate(1)) {
+		return epochStart.ErrInvalidInflationRate
```
Comment (iulianpascalau) [inline epochStart/metachain/economics.go:435]: suggestion to wrap all these errors with the compared value & limits. It will be easier to debug.

Comment (AdoAdoAdo) [reply epochStart/metachain/economics.go]: done

File: epochStart/metachain/economics.go (hunk 1 above)
Comment (AdoAdoAdo) [reply epochStart/metachain/economics.go]: Let's discuss and implement in a follow up PR.

**Review by iulianpascalau — DISMISSED** [review]

**Review by raduchis — APPROVED** [review]

**Review by iulianpascalau — DISMISSED** [review]

**Review by raduchis — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by stefanandy — APPROVED** [review]
System Test Passed

**Pull request merged**